accounts/abi/bind/backends/simulated.go

```diff
@@ -398,39 +404,67 @@ func (b *SimulatedBackend) EstimateGas(ctx context.Context, call ethereum.CallMs
 	cap = hi
 
 	// Create a helper to check if a gas allowance results in an executable transaction
-	executable := func(gas uint64) bool {
+	executable := func(gas uint64) (bool, *core.ExecutionResult, error) {
 		call.Gas = gas
 
 		snapshot := b.pendingState.Snapshot()
-		_, _, failed, err := b.callContract(ctx, call, b.pendingBlock, b.pendingState)
+		res, err := b.callContract(ctx, call, b.pendingBlock, b.pendingState)
 		b.pendingState.RevertToSnapshot(snapshot)
 
-		if err != nil || failed {
-			return false
+		if err != nil {
+			if err == core.ErrIntrinsicGas {
+				return true, nil, nil // Special case, raise gas limit
+			}
+			return true, nil, err // Bail out
 		}
-		return true
+		return res.Failed(), res, nil
 	}
 	// Execute the binary search and hone in on an executable gas limit
 	for lo+1 < hi {
 		mid := (hi + lo) / 2
-		if !executable(mid) {
+		failed, _, err := executable(mid)
+
+		// If the error is not nil(consensus error), it means the provided message
+		// call or transaction will never be accepted no matter how much gas it is
+		// assigned. Return the error directly, don't struggle any more
```

Comment:
This isn't strictly true.  A transaction may have a line like `require(msg.gas < 1,000,000, "Not enough gas");` or similar.  We may not care about this class of failures (where the failure is dependent on gas, timestamp, block number, etc.) but there may be value in adding clarity to the comment.

Reply:
@MicahZoltu It might help if we could collect a small pool of solidity code to add to tests and the desired behavior in those cases.

Code that depends on the gas counter is a bit hard to handle correctly. e.g.

```
require(msg.gas < 1,000,000, "Not enough gas")
require(msg.gas > 4,000,000, "Too much gas")
```

^ cannot be binary searched, so the estimator cannot ever properly pick a correct number. It might be a valid corner case though to support estimating these kinds of txs as long as they can be binary searched.

Reply:
Feel free to open a new issue with a code that fails estimation and we can add some tweaks.

Reply:
Ah sorry, I don't think I actually want support for these edge cases, it feels not worth the effort.  I was merely stating that the comment in the code isn't strictly correct and for clarity to future readers it may be valuable to mention something like "we don't care about these edge cases at the moment" in the comment.

Reply:
@MicahZoltu Actually the scenario you described is handled. Since all `revert` will lead to an internal evm error which is wrapped by `ExecutionResult`. However the error here is another one refers to some consensus issue(e.g. the nonce is not correct), so in this case no matter how many gases is assigned, the transaction will never be accpeted.